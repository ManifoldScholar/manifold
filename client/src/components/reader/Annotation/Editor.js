import React, { PureComponent, PropTypes } from 'react';
import { Link } from 'react-router';
import { connect } from 'react-redux';
import classNames from 'classnames';
import { isPromise } from 'utils/promise';
import { Form as GlobalForm } from 'components/global';

export default class AnnotationSelectionEditor extends PureComponent {

  static displayName = "Annotation.Selection.Editor";

  static propTypes = {
    id: PropTypes.string,
    body: PropTypes.string,
    isPrivate: PropTypes.bool,
    subject: PropTypes.string.isRequired,
    startNode: PropTypes.string.isRequired,
    startChar: PropTypes.number.isRequired,
    endNode: PropTypes.string.isRequired,
    endChar: PropTypes.number.isRequired,
    cancel: PropTypes.func.isRequired,
    saveHandler: PropTypes.func.isRequired,
    closeOnSave: PropTypes.bool,
    addsTo: PropTypes.string
  }

  constructor(props) {
    super(props);

    this.handleBodyChange = this.handleBodyChange.bind(this);
    this.handlePrivacyChange = this.handlePrivacyChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleCancel = this.handleCancel.bind(this);

    this.state = {
      body: "",
      isPrivate: false,
      errors: []
    };

    if (props.body) this.state.body = props.body;
    if (props.private) this.state.isPrivate = props.private;
  }

  componentDidMount() {
    this.ci.focus();
  }

  handleSubmit(event) {
    event.preventDefault();
    const { subject, startNode, startChar, endNode, endChar } = this.props;
    const { body, isPrivate } = this.state;
    const annotation = {
      subject,
      startNode,
      startChar,
      endNode,
      endChar,
      body,
      // eslint-disable-next-line quote-props
      "private": isPrivate,
      format: "annotation"
    };
    if (this.props.id) annotation.id = this.props.id;
    const options = { closeOnSave: this.props.closeOnSave };
    if (this.props.addsTo) options.addsTo = this.props.addsTo;
    const promise = this.props.saveHandler(annotation, options);
    if (isPromise(promise)) {
      promise.then(() => {
        this.props.cancel();
      }, (response) => {
        this.handleErrors(response.body.errors);
      });
    }
  }

  handleBodyChange(event) {
    this.setState({ body: event.target.value });
  }

  handleCancel(event) {
    event.preventDefault();
    if (this.props.cancel) {
      this.props.cancel(event);
    }
  }

  handlePrivacyChange(event) {
    const value = !this.state.isPrivate;
    this.setState({ isPrivate: value });
  }

  handleErrors(errors) {
    this.setState({ errors });
  }

  render() {

    const checkClass = classNames(
      'form-toggle',
      'checkbox',
      { checked: this.state.isPrivate }
    );

    return (
      <div className="annotation-editor">
        <form onSubmit={this.handleSubmit}>
          <GlobalForm.Errorable
            name="attributes[body]"
            errors={this.state.errors}
          >
            <textarea
              ref={(ci) => { this.ci = ci; }}
              style={{ width: "100%" }}
              placeholder={'Annotate this passage...'}
              onChange={this.handleBodyChange}
              value={this.state.body}
            />
          </GlobalForm.Errorable>
          <div className="utility">
            <div className="form-input">
              <label className={checkClass} >
                <input
                  type="checkbox"
                  name="isPrivate"
                  value="1"
                  checked={this.state.isPrivate}
                  onChange={this.handlePrivacyChange}
                />
                <span className="toggle-indicator">
                <i className="manicon manicon-check-bold"></i>
              </span>
                <span className="toggle-label">This Annotation is Private</span>
              </label>
            </div>
            <div className="buttons">
              <button
                onClick={this.handleCancel}
                className="button-primary dull"
              >
                Cancel
              </button>
              <button
                className="button-secondary"
                disabled={!this.state.body}
              >
                Save
              </button>
            </div>
          </div>

        </form>
      </div>
    );
  }

}
