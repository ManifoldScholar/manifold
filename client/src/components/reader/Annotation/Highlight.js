import React, { PureComponent } from "react";
import { Utility } from "components/frontend";
import PropTypes from "prop-types";

export default class HighlightDetail extends PureComponent {
  static displayName = "Annotation.Highlight";

  static propTypes = {
    annotation: PropTypes.object.isRequired,
    deleteHandler: PropTypes.func,
    visitHandler: PropTypes.func
  };

  handleDelete = event => {
    if (event) event.preventDefault();
    this.props.deleteHandler(this.props.annotation);
  };

  handleVisitHighlight = event => {
    event.preventDefault();
    this.props.visitHandler(this.props.annotation);
  };

  render() {
    const annotation = this.props.annotation;
    return (
      <div className="annotation-highlight-detail">
        <span className="annotation-selection">
          {annotation.attributes.subject}
        </span>

        <nav className="utility">
          <ul>
            {this.props.visitHandler ? (
              <li>
                <button onClick={this.handleVisitHighlight}>
                  {"View In Text"}
                </button>
              </li>
            ) : null}
            {this.props.deleteHandler &&
            annotation.attributes.canUpdateObject ? (
              <li>
                <Utility.ConfirmableButton
                  label="Delete"
                  confirmHandler={this.handleDelete}
                />
              </li>
            ) : null}
          </ul>
        </nav>
      </div>
    );
  }
}

//
