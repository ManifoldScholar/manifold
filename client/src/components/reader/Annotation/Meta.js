import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { FormattedDate } from "components/global";
import classNames from "classnames";

export default class AnnotationDetail extends PureComponent {
  static displayName = "Annotation.Meta";

  static propTypes = {
    creator: PropTypes.object.isRequired,
    annotation: PropTypes.object.isRequired,
    showAnnotationLabel: PropTypes.bool,
    subject: PropTypes.string
  };

  static defaultProps = {
    showAnnotationLabel: false
  };

  subtitle() {
    if (!this.props.subject) return this.dateSubtitle();
    return this.subjectSubtitle();
  }

  name() {
    const { creator, annotation } = this.props;
    const isCreator = annotation.attributes.abilities.creator;
    let name = creator.attributes.fullName;
    if (isCreator) name = "Me";
    return (
      <h4 className="author-name">
        {name}
      </h4>
    );
  }

  subjectSubtitle() {
    const { subject } = this.props;
    return (
      <div className="subtitle">
        {subject} {this.dateSubtitle()}
      </div>
    );
  }

  dateSubtitle() {
    const { annotation } = this.props;
    return (
      <datetime>
        <FormattedDate
          format="distanceInWords"
          date={annotation.attributes.createdAt}
        />{" "}
        ago
      </datetime>
    );
  }

  render() {
    const { creator, annotation } = this.props;
    const avatarClass = classNames({
      "author-avatar": true,
      dull: !creator.attributes.isCurrentUser
    });

    return (
      <section className="annotation-meta">
        {/* NB: Empty div required for flex-positioning of private/author marker */}
        <div>
          <figure className={avatarClass}>
            {creator.attributes.avatarStyles.smallSquare
              ? <div
                  className="image"
                  style={{
                    backgroundImage: `url(${creator.attributes.avatarStyles
                      .smallSquare})`
                  }}
                >
                  <span className="screen-reader-text">
                    Profile image for {creator.attributes.fullName}
                  </span>
                </div>
              : <div className="no-image">
                  <i className="manicon manicon-person" />
                </div>}
          </figure>
          {this.name()}
          {this.subtitle()}
        </div>
        {annotation.attributes.private
          ? <div className="marker secondary">
              {"Private"}
            </div>
          : null}
        {this.props.showAnnotationLabel
          ? <div className="marker tertiary">
              {"Annotation"}
            </div>
          : null}
      </section>
    );
  }
}
