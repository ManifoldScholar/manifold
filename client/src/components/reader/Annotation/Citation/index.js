import Editor from './Editor';
import Wrapper from './Wrapper';

export default {
  Editor,
  Wrapper
};
