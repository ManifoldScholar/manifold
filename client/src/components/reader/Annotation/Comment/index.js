import Detail from './Detail';
import Editor from './Editor';

export default {
  Detail,
  Editor
}