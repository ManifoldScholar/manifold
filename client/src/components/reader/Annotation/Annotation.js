import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { Helper } from "components/global";
import { Utility } from "components/frontend";
import Editor from "./Editor";
import Meta from "./Meta";
import { Comment as CommentContainer } from "containers/global";
import classNames from "classnames";
import HigherOrder from "containers/global/HigherOrder";

export default class AnnotationDetail extends PureComponent {
  static displayName = "Annotation.Detail";

  static propTypes = {
    creator: PropTypes.object.isRequired,
    annotation: PropTypes.object.isRequired,
    saveHandler: PropTypes.func,
    deleteHandler: PropTypes.func,
    showLogin: PropTypes.func,
    includeComments: PropTypes.bool.isRequired
  };

  static defaultProps = {
    includeComments: true
  };

  constructor(props) {
    super(props);

    this.state = {
      action: null
    };

    this.startReply = this.startReply.bind(this);
    this.startEdit = this.startEdit.bind(this);
    this.stopAction = this.stopAction.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
  }

  startReply() {
    this.setState({
      action: "replying"
    });
  }

  startEdit() {
    this.setState({
      action: "editing"
    });
  }

  stopAction() {
    this.setState({
      action: null
    });
  }

  handleDelete = event => {
    if (event) event.preventDefault();
    this.props.deleteHandler(this.props.annotation);
  };

  render() {
    const replyButtonClass = classNames({
      active: this.state.action === "replying"
    });
    const editButtonClass = classNames({
      active: this.state.action === "editing"
    });

    const creator = this.props.creator;
    const annotation = this.props.annotation;

    return (
      <li className="annotation-annotation">
        <Meta annotation={annotation} creator={creator} />
        {this.state.action === "editing" ? (
          <Editor
            id={annotation.id}
            body={annotation.attributes.body}
            private={annotation.attributes.private}
            subject={annotation.attributes.subject}
            startNode={annotation.attributes.startNode}
            startChar={annotation.attributes.startChar}
            endNode={annotation.attributes.endNode}
            endChar={annotation.attributes.endChar}
            saveHandler={this.props.saveHandler}
            cancel={this.stopAction}
          />
        ) : (
          <div>
            <section className="body">
              <Helper.SimpleFormat text={annotation.attributes.body} />
            </section>
            <HigherOrder.Authorize kind={"any"}>
              <nav className="utility">
                <ul>
                  {this.props.includeComments ? (
                    <li>
                      <button
                        className={replyButtonClass}
                        onClick={this.startReply}
                      >
                        {"Reply"}
                      </button>
                    </li>
                  ) : null}
                  <HigherOrder.Authorize entity={annotation} ability={"update"}>
                    {this.props.saveHandler ? (
                      <li>
                        <button
                          className={editButtonClass}
                          onClick={this.startEdit}
                        >
                          {"Edit"}
                        </button>
                      </li>
                    ) : null}
                  </HigherOrder.Authorize>
                  <HigherOrder.Authorize entity={annotation} ability={"delete"}>
                    {this.props.deleteHandler ? (
                      <li>
                        <Utility.ConfirmableButton
                          label="Delete"
                          confirmHandler={this.handleDelete}
                        />
                      </li>
                    ) : null}
                  </HigherOrder.Authorize>
                </ul>
                {this.state.action === "replying" ? (
                  <CommentContainer.Editor
                    subject={annotation}
                    cancel={this.stopAction}
                  />
                ) : null}
              </nav>
            </HigherOrder.Authorize>
            <HigherOrder.Authorize kind="unauthenticated">
              <nav className="utility">
                <ul>
                  <li>
                    <button onClick={this.props.showLogin}>
                      {"Login to reply"}
                    </button>
                  </li>
                </ul>
              </nav>
            </HigherOrder.Authorize>
          </div>
        )}
        {this.props.includeComments ? (
          <CommentContainer.Thread subject={annotation} />
        ) : null}
      </li>
    );
  }
}

//
