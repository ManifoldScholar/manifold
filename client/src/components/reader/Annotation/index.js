import Annotation from "./Annotation";
import Editor from "./Editor";
import Highlight from "./Highlight";
import Popup from "./Popup";
import Selection from "./Selection";
import Share from "./Share";

export default {
  Annotation,
  Editor,
  Highlight,
  Popup,
  Selection,
  Share
};
