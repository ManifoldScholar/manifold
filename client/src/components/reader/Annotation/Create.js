import React, { PureComponent, PropTypes } from 'react';
import { Annotation } from 'components/reader';

export default class AnnotationCreate extends PureComponent {

  static displayName = "Annotation.Create";

  static propTypes = {
    annotation: PropTypes.object,
    selection: PropTypes.object,
    closeDrawer: PropTypes.func
  }

  render() {
    return (
      <Annotation.Selection.Wrapper {...this.props} annotating />
    );
  }

}
