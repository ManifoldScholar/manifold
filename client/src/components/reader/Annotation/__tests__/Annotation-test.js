import React from 'react';
import { Provider } from 'react-redux';
import Annotation from '../Annotation';
import renderer from 'react-test-renderer';
import build from 'test/fixtures/build';
import { wrapWithRouter, renderWithRouter } from 'test/helpers/routing';

describe("Reader.Annotation.Annotation component", () => {

  const creator = build.entity.user("1");
  const annotation = build.entity.annotation("2");
  const store = build.store();

  const root = (
    wrapWithRouter(
      <Provider store={store} >
        <Annotation
          creator={creator}
          annotation={annotation}
        />
      </Provider>
    )
  );

  it('renders correctly', () => {
    const component = renderer.create(root);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
