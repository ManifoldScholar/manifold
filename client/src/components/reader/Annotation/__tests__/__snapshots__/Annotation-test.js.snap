// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reader.Annotation.Annotation component renders correctly 1`] = `
<li
  className="annotation-annotation"
>
  <section
    className="meta"
  >
    <div>
      <figure
        className="author-avatar"
      >
        <div
          className="no-image"
        >
          <i
            className="manicon manicon-person"
          />
        </div>
      </figure>
      <h4
        className="author-name"
      >
        Rowan Ida
      </h4>
      <datetime>
         
        ago
      </datetime>
    </div>
  </section>
  <div>
    <section
      className="body"
    >
      <div>
        <p>
          <span>
            Hands on your boxes, turn 'em up like seven notches. Your Magnavoxes amplify my super conscious.
          </span>
        </p>
      </div>
    </section>
    <nav
      className="utility"
    >
      <ul>
        <li>
          <button
            onClick={undefined}
          >
            Login to reply
          </button>
        </li>
      </ul>
    </nav>
  </div>
  <div>
    <button
      className="comment-more"
    >
      <i
        className="manicon manicon-word-bubble-multiple"
      />
      See all 36 replies
    </button>
  </div>
</li>
`;
